### PR TITLE
TINYGL: Avoid calling memset_{s,l} when possible.

### PR DESCRIPTION
**Pull request by vpelletier**
It is possible to convert calls to memset_{s,l} into call to memset when
all bytes of the depth or color value are the same (which is the likely
case), improving buffer clearing speed.

For reference, I developped this to allow adding support for 32bits rendering in TinyGL, but is a stand-alone change.

### CODE REVIEW

File: graphics/tinygl/zbuffer.cpp [hunk 1]
```diff
@@ -141,14 +141,28 @@ void FrameBuffer::delOffscreenBuffer(Buffer *buf) {
 
 void FrameBuffer::clear(int clearZ, int z, int clearColor, int r, int g, int b) {
 	if (clearZ) {
-		memset_l(this->_zbuf, z, this->xsize * this->ysize);
+		const uint8 *zc = (const uint8 *) &z;
+		unsigned int i;
+		for (i = 1; i < sizeof(z) && zc[0] == zc[i]; i++);
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: please use { ; }

File: graphics/tinygl/zbuffer.cpp [hunk 2]
```diff
@@ -141,14 +141,28 @@ void FrameBuffer::delOffscreenBuffer(Buffer *buf) {
...
 	}
 	if (clearColor) {
 		byte *pp = this->pbuf.getRawBuffer();
 		uint32 color = this->cmode.RGBToColor(r, g, b);
-		for (int y = 0; y < this->ysize; y++) {
-			memset_s(pp, color, this->xsize);
-			pp = pp + this->linesize;
+		const uint8 *colorc = (uint8 *) &color;
+		unsigned int i;
+		for (i = 1; i < sizeof(color) && colorc[0] == colorc[i]; i++);
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: please use { ; }

File: graphics/tinygl/zbuffer.cpp [hunk 3]
```diff
@@ -162,9 +176,18 @@ void FrameBuffer::clearRegion(int x, int y, int w, int h, int clearZ, int z, int
 	if (clearColor) {
 		byte *pp = this->pbuf.getRawBuffer() + y * this->linesize;
 		uint32 color = this->cmode.RGBToColor(r, g, b);
-		for (int row = y; row < y + h; row++) {
-			memset_s(pp + x * this->pixelbytes, color, w);
-			pp = pp + this->linesize;
+		const uint8 *colorc = (uint8 *) &color;
+		unsigned int i;
+		for (i = 1; i < sizeof(color) && colorc[0] == colorc[i]; i++);
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: please use { ; }

**Comment by vpelletier** [comment]
Thanks, latest push fix this.
It also contains 2 other improvements:
- also avoid memset_{s,l} for `_zbuf` in `clearRegion`
- use `sizeof(_zbuf)` instead of `sizeof(z)` as we are basically guessing the intended argument byte size, so I think referring to _zbuf makes the intent clearer

File: graphics/tinygl/zbuffer.cpp [hunk 4]
```diff
@@ -141,30 +141,65 @@ void FrameBuffer::delOffscreenBuffer(Buffer *buf) {
...
 		}
 	}
 }
 
 void FrameBuffer::clearRegion(int x, int y, int w, int h, int clearZ, int z, int clearColor, int r, int g, int b) {
 	if (clearZ) {
-		for (int row = y; row < y + h; row++) {
-			memset_l(this->_zbuf + x + (row * this->xsize), z, w);
+		unsigned int *zbuf = this->_zbuf + (y * this->xsize);
+		const uint8 *zc = (const uint8 *) &z;
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: remove space before &z;

File: graphics/tinygl/zbuffer.cpp [hunk 5]
```diff
@@ -141,30 +141,65 @@ void FrameBuffer::delOffscreenBuffer(Buffer *buf) {
...
+		unsigned int i;
+		for (i = 1; i < sizeof(z) && zc[0] == zc[i]; i++) { ; }
+		if (i == sizeof(z)) {
+			while (h--) {
+				memset(zbuf + x, zc[0], sizeof(*zbuf) * w);
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: Add space between *zbuf

File: graphics/tinygl/zbuffer.cpp [hunk 6]
```diff
@@ -141,30 +141,65 @@ void FrameBuffer::delOffscreenBuffer(Buffer *buf) {
 
 void FrameBuffer::clear(int clearZ, int z, int clearColor, int r, int g, int b) {
 	if (clearZ) {
-		memset_l(this->_zbuf, z, this->xsize * this->ysize);
+		const uint8 *zc = (const uint8 *) &z;
+		unsigned int i;
+		for (i = 1; i < sizeof(z) && zc[0] == zc[i]; i++) { ; }
+		if (i == sizeof(z)) {
+			memset(this->_zbuf, zc[0], sizeof(*this->_zbuf) * this->xsize * this->ysize);
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: Add space between * and this->_zbuf

File: graphics/tinygl/zbuffer.cpp [hunk 7]
```diff
@@ -141,30 +141,65 @@ void FrameBuffer::delOffscreenBuffer(Buffer *buf) {
 
 void FrameBuffer::clear(int clearZ, int z, int clearColor, int r, int g, int b) {
 	if (clearZ) {
-		memset_l(this->_zbuf, z, this->xsize * this->ysize);
+		const uint8 *zc = (const uint8 *) &z;
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: remove space before &z;

File: graphics/tinygl/zbuffer.cpp [hunk 8]
```diff
@@ -141,30 +141,65 @@ void FrameBuffer::delOffscreenBuffer(Buffer *buf) {
...
+			memset_l(this->_zbuf, z, this->xsize * this->ysize);
+		}
 	}
 	if (clearColor) {
 		byte *pp = this->pbuf.getRawBuffer();
 		uint32 color = this->cmode.RGBToColor(r, g, b);
-		for (int y = 0; y < this->ysize; y++) {
-			memset_s(pp, color, this->xsize);
-			pp = pp + this->linesize;
+		const uint8 *colorc = (uint8 *) &color;
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: remove space before &z;

File: graphics/tinygl/zbuffer.cpp [hunk 9]
```diff
@@ -141,30 +141,65 @@ void FrameBuffer::delOffscreenBuffer(Buffer *buf) {
...
+			}
 		}
 	}
 	if (clearColor) {
 		byte *pp = this->pbuf.getRawBuffer() + y * this->linesize;
 		uint32 color = this->cmode.RGBToColor(r, g, b);
-		for (int row = y; row < y + h; row++) {
-			memset_s(pp + x * this->pixelbytes, color, w);
-			pp = pp + this->linesize;
+		const uint8 *colorc = (uint8 *) &color;
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: remove space before &color;

**Comment by vpelletier** [comment]
Thanks, fixed & rebased over latest master.

File: graphics/tinygl/zbuffer.cpp [hunk 10]
```diff
@@ -139,30 +139,65 @@ void FrameBuffer::delOffscreenBuffer(Buffer *buf) {
 
 void FrameBuffer::clear(int clearZ, int z, int clearColor, int r, int g, int b) {
```
Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: Maybe you could add a little dexription how function works. It's not so clear.

Comment (vpelletier) [inline graphics/tinygl/zbuffer.cpp]: Is it clearer now ? Or is it the way I use the `for` loop which is surprising ?

Also, thanks for making me take a 3rd look at this code, as I could simplify `clear` a bit and discovered `clearRegion` was broken.

Comment (aquadran) [inline graphics/tinygl/zbuffer.cpp]: I think is fine now

**Pull request merged**